internal/invocation/invocation.go

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package invocation
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	pubcluster "github.com/hazelcast/hazelcast-go-client/cluster"
+	"github.com/hazelcast/hazelcast-go-client/internal/hzerror"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto"
+)
+
+type Result interface {
+	Get() (*proto.ClientMessage, error)
+	GetWithTimeout(duration time.Duration) (*proto.ClientMessage, error)
+}
+
+type Invocation interface {
+	Complete(message *proto.ClientMessage)
+	Completed() bool
+	EventHandler() proto.ClientMessageHandler
+	Get() (*proto.ClientMessage, error)
+	GetContext(ctx context.Context) (*proto.ClientMessage, error)
+	GetWithTimeout(duration time.Duration) (*proto.ClientMessage, error)
+	PartitionID() int32
+	Request() *proto.ClientMessage
+	Address() *pubcluster.AddressImpl
+}
+
+type Impl struct {
+	request         *proto.ClientMessage
+	response        chan *proto.ClientMessage
+	completed       int32
+	boundConnection *Impl
+	address         *pubcluster.AddressImpl
+	partitionID     int32
+	//sentConnection  atomic.Value
+	eventHandler func(clientMessage *proto.ClientMessage)
+	deadline     time.Time
+}
+
+func NewImpl(clientMessage *proto.ClientMessage, partitionID int32, address *pubcluster.AddressImpl, timeout time.Duration) *Impl {
+	return &Impl{
+		partitionID: partitionID,
+		address:     address,
+		request:     clientMessage,
+		response:    make(chan *proto.ClientMessage, 1),
+		deadline:    time.Now().Add(timeout),
+	}
+}
+
+func (i *Impl) Complete(message *proto.ClientMessage) {
+	if atomic.CompareAndSwapInt32(&i.completed, 0, 1) {
+		i.response <- message
+	}
+}
+
+func (i *Impl) Completed() bool {
+	return i.completed != 0
+}
+
+func (i *Impl) EventHandler() proto.ClientMessageHandler {
+	return i.eventHandler
+}
+
+func (i *Impl) Get() (*proto.ClientMessage, error) {
+	response := <-i.response
+	return i.unwrapResponse(response)
+}
+
+func (i *Impl) GetContext(ctx context.Context) (*proto.ClientMessage, error) {
```

Comment:
This name sounds a lot like a simple getter.
Can we rename it to GetWithContext ?

Reply:
Updated at: https://github.com/yuce/hazelcast-go-client/commit/925534a086ef8ae094679c0ee9afcd3ee8d51e13